beam/src/Rhyolite/Task/Beam.hs

```diff
@@ -0,0 +1,20 @@
+{-# Language FunctionalDependencies #-}
+{-# Language KindSignatures #-}
+-- | Contains definitions for tasks, and tables that have tasks.
+module Rhyolite.Task.Beam where
+
+import Data.Text (Text)
+
+import Database.Beam
+
+-- | Data Family for tables that have tasks embedded in them.
+-- Specifically, this means tables that have the following 2 fields (columns)
+-- 1. A result field, containg an optional value
+-- 2. A checkout out by field, containg the name of the worker that checked out this task
+-- For unclaimed tasks, both these fields are null (Nothing).
+-- When a task is checked out by a worker, the checkout out field is set to the worker's name. Result is still null.
+-- When a task is completed, the result field is filled with the result, and the checked out field is set to null.
+-- The type of result is dependent upon the task, hence the functional dependency between the table and the result type.
+class HasTasks (table :: (* -> *) -> *) a | table -> a where
```

Comment:
In the groundhog instance of this library we were very limited in how we could structure the Task abstraction. However with Beam we have more freedom. In particular, you can simply nest `Beamable` types and have them be instantiated into columns when the top-level type is included in a database.

I would like you to try the following approach and see if it works: Create a Beamable `Task` type (so it has a type constructor parameter) with two parameters, one for the payload of the Task which is passed to the worker and one for the result of the Task which is reported by the worker. So the Task type will look something like

```
data Task f i o = Task
  { _task_payload :: C f i
  , _task_result :: C f (Maybe o)
  , _task_checkedOutBy :: C f (Maybe Text)
  }
 ```
 
Use this `lens` class which demands more structure than `HasTasks` instead (because a lens has more structure than a function): https://hackage.haskell.org/package/lens-5.1/docs/Control-Lens-TH.html#v:makeClassy

You should then be able to implement `taskWorker` for any table type that has a Task field (Task will be expanded into columns by Beam).

There will be some technical complications around lenses and table entities in beam, but they are supported so lets see if we can take this idea and make it work.

Reply:
This idea also sounds good to me. If we can avoid having to write instances of an extra class to pick out which columns mean the things we want, that's probably much nicer to use.

Reply:
@madeline-os I have made the changes, but I could not get the code to compile. Will look at it on Monday.

@cgibbard I thought that any table having 2 columns for tasks will automatically fit our class. Even if I use @madeline-os's suggestion, we still have to implement a class (instead of creating our own class, we will be using one that `lens` library created for us). Let me know if this understanding is incorrect.

Reply:
The `lens` class can be derived.